contracts/src/lib/SupplySourceHelper.sol

```diff
@@ -101,13 +101,31 @@ library SupplySourceHelper {
             // Use the optimized path to send value along with the call.
             (success, ret) = functionCallWithValue({target: target, data: data, value: value});
         } else if (supplySource.kind == SupplyKind.ERC20) {
-            // Transfer the tokens first, _then_ perform the call.
-            (success, ret) = ierc20Transfer(supplySource, target, value);
-            if (success) {
-                // Perform the call only if the ERC20 was successful.
-                (success, ret) = functionCallWithValue(target, data, 0);
-            } else {
-                return (success, ret);
+            (success, ret) = functionCallWithERC20Value({supplySource: supplySource, target: target, data: data, value: value});
+        }
+        return (success, ret);
+    }
+
+    /// @dev Performs the function call with ERC20 value atomically
+    function functionCallWithERC20Value(
+        SupplySource memory supplySource,
+        address target,
+        bytes memory data,
+        uint256 value
+    ) internal returns (bool success, bytes memory ret) {
+        // Transfer the tokens first, _then_ perform the call.
+        (success, ret) = ierc20Transfer(supplySource, target, value);
+
+        if (success) {
+            // Perform the call only if the ERC20 was successful.
+            (success, ret) = functionCallWithValue(target, data, 0);
+        }
+        
+        if (!success) {
+            // following the implementation of `openzeppelin-contracts/utils/Address.sol`
+            assembly {
+                let returndata_size := mload(ret)
+                revert(add(32, ret), returndata_size)
             }
```

Comment:
Could you comment why the `add` opcode? There may be comments in the OZ library, but let's avoid reader indirection.

Reply:
Ok, I will add comments on the code.

Reply:
This is the full function from OpenZeppelin. We need to handle the case where there is no return data. The `add` is needed to unpack the variable-size return value (first Ethereum word is the size).

```solidity
function _revert(bytes memory returndata) private pure {
    // Look for revert reason and bubble it up if present
    if (returndata.length > 0) {
        // The easiest way to bubble the revert reason is using memory via assembly
        /// @solidity memory-safe-assembly
        assembly {
            let returndata_size := mload(returndata)
            revert(add(32, returndata), returndata_size)
        }
    } else {
        revert FailedInnerCall();
    }
}
```

Reply:
From my testing, if the `ret` is empty, calling `revert(add(32, returndata), returndata_size)` is the same as revert with no data. I think openzeppelin might be emphasising `FailedInnerCall`? But for us, if the contract returns empty, we can just return empty. Even if we dont, after some parsing of the error bytes, users will still get an empty ret.

Reply:
From my testing, if the `ret` is empty, calling `revert(add(32, returndata), returndata_size)` is the same as revert with no data. I think openzeppelin might be emphasising `FailedInnerCall`? But for us, if the contract returns empty, we can just return empty. Even if we dont, after some parsing of the error bytes, users will still get an empty ret.

Reply:
@cryptoAtwill I suspect you're seeing this because mloading the return data pointer will return the size, which will be 0. You then revert providing an offset equal to the pointer+32, **but with size 0**. This is probably OK, but I'd rather play it safe and do the check anyway in high level Solidity.